### ISSUE TITLE
Wishlist: automatically split pointer tables

### ISSUE THREAD

**Issue by clbr** [cc65/cc65#350]
Over at nesdev, there's been talk about how it would be useful for ca65 to automatically split 16-bit data tables, perhaps with a macro. It would also be useful in C, when the compiler knows for sure it can do so (with the table declared static const).

Example:

```c
void foo(void);
void bar(void);

// Function pointers in this example, but any pointers or 16-bit data would apply
static void (* const funcs[2])(void) = {
	foo,
	bar
};

void call(unsigned char what) {
	funcs[what]();
}
```

Current:
```asm
;
; File generated by cc65 v 2.15 - Git 24c2da9
;
	.fopt		compiler,"cc65 v 2.15 - Git 24c2da9"
	.setcpu		"6502"
	.smart		on
	.autoimport	on
	.case		on
	.debuginfo	off
	.importzp	sp, sreg, regsave, regbank
	.importzp	tmp1, tmp2, tmp3, tmp4, ptr1, ptr2, ptr3, ptr4
	.macpack	longbranch
	.import		_foo
	.import		_bar
	.export		_call

.segment	"RODATA"

_funcs:
	.addr	_foo
	.addr	_bar

; ---------------------------------------------------------------
; void __near__ call (unsigned char)
; ---------------------------------------------------------------

.segment	"CODE"

.proc	_call: near

.segment	"CODE"

;
; void call(unsigned char what) {
;
	jsr     pusha
;
; funcs[what]();
;
	ldx     #$00
	lda     (sp,x)
	asl     a
	bcc     L0008
	inx
	clc
L0008:	adc     #<(_funcs)
	sta     ptr1
	txa
	adc     #>(_funcs)
	sta     ptr1+1
	ldy     #$01
	lda     (ptr1),y
	tax
	dey
	lda     (ptr1),y
	jsr     callax
;
; }
;
	jmp     incsp1

.endproc
```

Desired:
```asm
;
; File generated by cc65 v 2.15 - Git 24c2da9
;
	.fopt		compiler,"cc65 v 2.15 - Git 24c2da9"
	.setcpu		"6502"
	.smart		on
	.autoimport	on
	.case		on
	.debuginfo	off
	.importzp	sp, sreg, regsave, regbank
	.importzp	tmp1, tmp2, tmp3, tmp4, ptr1, ptr2, ptr3, ptr4
	.macpack	longbranch
	.import		_foo
	.import		_bar
	.export		_call

.segment	"RODATA"

_funcsL:
	.byte	<(_foo)
	.byte	<(_bar)
_funcsH:
	.byte	>(_foo)
	.byte	>(_bar)

; ---------------------------------------------------------------
; void __near__ call (unsigned char)
; ---------------------------------------------------------------

.segment	"CODE"

.proc	_call: near

.segment	"CODE"

;
; void call(unsigned char what) {
;
	tay
;
; funcs[what]();
;
	lda	_funcsH,y
	tax
	lda	_funcsL,y

	jsr     callax
;
; }
;
	rts

.endproc
```

This sort of transformation can greatly speed up the code, it's a PITA to do manually, and yet it's something the compiler could do easily.

**Comment by IrgendwerA8** [comment]
A compromise would be an enhanced support of constant expression. Yes, it a tedious work to do the above manually but it won't work even now without runtime loss:

```
extern void funcA();
extern void funcB();

unsigned char funcLow[] = { (unsigned int)funcA & 0xFF, (unsigned int)funcB & 0xFF };
unsigned char funcHigh[] = { (unsigned int)funcA >> 8, (unsigned int)funcB >> 8 };
```

are currently not constant expressions, which is bad on a platform with quite restricted memory and performance. AFAIK a change wouldn't violate the standard.

BTW: Something like
`unsigned char funcLow[] = { (unsigned char)funcA, (unsigned char)funcB};`
doesn't help. It is currently also non-const (otherwise you could fiddle with the address and get also the high-byte this way). Currently only const additions result in constness and are ok. (adr + 1 works, adr -1 not,  but adr + -1 is ok.)

If there would be at least a change for this (also very valuable for data), a helper function using the tables could perform the call more efficiently.
To transform above definitions into the matching assembler ones is also a point standing high on my wish list.

**Comment by clbr** [comment]
Yeah, when I do that manually, I do it in the asm version. Doing it in C would be tedious^3.

**Comment by IrgendwerA8** [comment]
But that's what I like to have supported. In asm you have to import every function from the C side, for beginners sticking with C this isn't an option and you may only need for that purpose a s-file.

**Comment by An-S** [comment]
There are pseudo ops for that on assembly level: .lobytes and .hibytes.
From the compilers perspective there are some reason not to split tables. 
1.) If you consider a pointer to an int table entry, on which address it is meant to point to? The address of the lobyte or the address of the hibyte?
2.) If you typecast a pointer to an uint table entry to f.e (long int*) and assign it to an long int variable, what would be the result? The compiler could fetch the lobyte for the long int from the first part of the table, the mid byte from the second part, but what to do with the hi byte? From which part of the table should it be retrieved? In contrast, if the table is arranged in the usual way it is clear to fetch three succeeding bytes and transfer it to the long int.
3.) What would be the result of taking a pointer difference to a split table? What would be the result when casting two pointers to the split table of uint16_t values to (uint8_t *) and then calculating the difference? 
4.) How to fetch only the hibyte of a table entry of a split table? (uint8_t*)(adr)[1] would fail.

**Comment by clbr** [comment]
> There are pseudo ops for that on assembly level: .lobytes and .hibytes.

They require lots of manual wrangling still, for data in the normal
form. Long, long thread on nesdev.

> From the compilers perspective there are some reason not to split tables. 
> 1.) If you consider a pointer to an int table entry, on which address it is meant to point to? The address of the lobyte or the address of the hibyte?
> 2.) If you typecast a pointer to an uint table entry to f.e (long int*) and assign it to an long int variable, what would be the result? The compiler could fetch the lobyte for the long int from the first part of the table, the mid byte from the second part, but what to do with the hi byte? From which part of the table should it be retrieved? In contrast, if the table is arranged in the usual way it is clear to fetch three succeeding bytes and transfer it to the long int.
> 3.) What would be the result of taking a pointer difference to a split table? What would be the result when casting two pointers to the split table of uint16_t values to (uint8_t *) and then calculating the difference? 
> 4.) How to fetch only the hibyte of a table entry of a split table? (uint8_t*)(adr)[1] would fail. 

These are simply additional cases to block. If the code takes a pointer
to the table, do not split. It is not common to do so in NES code,
because cc65 pointers are slow - first copied to ZP pointers even if
already in ZP, then 16-bit added instead of the (ptr),y form.

**Comment by groessler** [comment]
On 09/19/16 09:31, clbr wrote:

> These are simply additional cases to block. If the code takes a pointer
> to the table, do not split.


You cannot do this. The pointer could be taken in a different source file.

**Comment by clbr** [comment]
In the first post, I specify this is only for "static const" tables. They must be readonly and only in that one file (yes, C allows casting away const, but the NES platform places RODATA into read-only memory, so it's guaranteed RO).

**Comment by groessler** [comment]
On 09/19/16 09:58, clbr wrote:

> In the first post, I specify this is only for "static const" tables. 
> They must be readonly and only in that one file (yes, C allows casting 
> away const, but the NES platform places RODATA into read-only memory, 
> so it's guaranteed RO).
>


Even if the table itself is "static const" a function inside this source 
file could return a pointer to this table to an outside function.

**Comment by clbr** [comment]
That would be caught by the "no taking pointers" check.

**Comment by groessler** [comment]
On 09/19/16 10:13, clbr wrote:

> That would be caught by the "no taking pointers" check.
>


Yep, sorry, this was a brain fart...

**Comment by mrdudz** [comment]
what about this instead? it would reduce the max. size of the array to 128 elements, but not break the pointer passing

```
.segment    "RODATA"

_funcs:
    .word   <(_foo)
    .word   <(_bar)
; ---------------------------------------------------------------
; void __near__ call (unsigned char)
; ---------------------------------------------------------------

.segment    "CODE"

.proc   _call: near

.segment    "CODE"

;
; void call(unsigned char what) {
;
    asl
    tay
;
; funcs[what]();
;
    ldx _funcs+1,y
    lda _funcs,y

    jsr     callax
;
; }
;
    rts

.endproc
```

**Comment by clbr** [comment]
I don't think that would work, because of the < (low byte) operator. But I get what you mean, the asl would be an improvement to the current generated code, but not as good as separated arrays would be.

**Comment by mrdudz** [comment]
> I don't think that would work, because of the < (low byte) operator. 
ooops :)
